Compute/storage separation and other fixes (#22098)
yugabyte/yugabyte-db: docs/content/preview/explore/going-beyond-sql/decoupling-compute-storage.md

## Why decouple?

Traditionally it has been necessary to keep Storage close to Compute in databases to achieve low latency. But with ultra-fast networks, CPUs, and memory in the cloud era, the separation of compute and storage resources in databases has advantages.

- **Scalability & Efficiency**: By decoupling you can scale each component independently and optimize resource utilization allowing you to add or remove computational power and storage capacity as needed. By isolating Compute and Storage resources, you can optimize the utilization of each component based on specific workloads. For instance, for growing storage-intensive workloads it is sufficient to scale just the storage layer while compute-intensive workloads like query parsing, sorting, aggregation, etc can be handled by scaling the compute layer.

> - **Heterogenous Hardware**: With decoupling, the flexibility of using different hardware configurations for the two layers becomes possible. For example, you can choose faster CPUs for the computational layer while selecting faster solid-state drives (SSDs) for the storage layer.

Reviewer: ```suggestion
- **Heterogenous Hardware**. Decoupling provides the flexibility to use different hardware configurations for the two layers. For example, you can choose faster CPUs for the computational layer, and select faster solid-state drives (SSDs) for the storage layer.
```

> - **Data placement flexibility**: This decoupling enables you to keep your data layer in certain geographies to adhere to data placement laws and at the same time allows you to specify different security restrictions for the storage and compute layer by placing them in different subnets.

Reviewer: ```suggestion
- **Data placement flexibility**. Decoupling enables you to keep your data layer in specific geographies to comply with data placement laws. It also allows you to specify different security restrictions for the storage and compute layer by placing them in different subnets.
```

Let us see how to accomplish this separation in YugabyteDB.

## Cluster setup

> Let's set up a local cluster with 6 nodes with IP addresses `127.0.0.[1-6]` and place them in zones `a-f`.

Reviewer: ```suggestion
Set up a local cluster with 6 nodes with IP addresses `127.0.0.[1-6]` and place them in zones `a-f`.
```

{{<setup/local numnodes=6 locations="aws.east.zone-a,aws.east.zone-b,aws.east.zone-c,aws.east.zone-d,aws.east.zone-e,aws.east.zone-f," >}}

> {{<note>}}For clarity, we are placing the nodes across 6 zones - a,b,c,d,e, and f. This will also ensure both your compute and storage nodes are spread across multiple zones.{{</note>}}

Reviewer: ```suggestion
{{<note>}}For clarity, we are placing the nodes across 6 zones, a through f. This also ensures both your compute and storage nodes are spread across multiple zones.{{</note>}}
```

## Standard behaviour

Typically, the application connects to all the nodes in the cluster. In our local cluster, the application will connect to the 6 nodes as shown in the following illustration.

> ![Basic cluster](/images/explore/decoupling-compute-storage-setup.png)

Reviewer: should the nodes be a-f rather than 1-6

> Each node has a YB-TServer service that is comprised of the [Query Layer(YQL)](../../../architecture/query-layer) and the [DocDB](../../../architecture/docdb)-based storage layer. Each node in the cluster does both query processing and storage of data. Now let us see how to divide the responsibilities of the compute-heavy query layer and the storage-heavy DocDB layer between the nodes.

Reviewer: ```suggestion
Each node has a YB-TServer service that is comprised of the [Query Layer (YQL)](../../../architecture/query-layer) and the [DocDB](../../../architecture/docdb)-based storage layer. Each node in the cluster does both query processing and storage of data. Now let's see how to divide the responsibilities of the compute-heavy query layer and the storage-heavy DocDB layer between the nodes.
```

## Separating storage

> You can use the [Tablespace](../tablespaces) feature in YugabyteDB to restrict the storage to certain zones. Let's say you want to store our data only in zones a,b, and c. For this, you need to create a tablespace limited to these zones. For e.g,

Reviewer: ```suggestion
You can use the [Tablespace](../tablespaces) feature in YugabyteDB to restrict the storage to certain zones. Suppose you want to store data only in zones a, b, and c. For this, you need to create a tablespace limited to these zones. For example:
```

```sql
CREATE TABLESPACE storage
  WITH (replica_placement='{"num_replicas": 3, "placement_blocks": [
    {"cloud":"aws","region":"east","zone":"zone-a","min_num_replicas":1},
    {"cloud":"aws","region":"east","zone":"zone-b","min_num_replicas":1},
    {"cloud":"aws","region":"east","zone":"zone-c","min_num_replicas":1}
]}');
```

> Now when you create your tables you have to attach them to the above tablespace. Say if you have a user table, then when you create the table attach it to the storage tablespace.

Reviewer: ```suggestion
Now when you create tables, you have to attach them to these tablespaces. Say you have a user table; when you create the table, attach it to the storage tablespace.
```

```sql
CREATE TABLE user (
    id INTEGER, name text
) TABLESPACE storage;
```

> This will automatically ensure that only zones a, b, and c. Now all the nodes located in these zones will be responsible for the storage of data. In our example as nodes 1, 2 and 3 are located in these zones, they will be responsible for data storage & nodes 4,5, and 6 outside these zones will not store any data.

Reviewer: ```suggestion
This automatically ensures that the data is only stored in zones a, b, and c. Now all the nodes located in these zones will be responsible for the storage of data. Nodes in zones d, e, and f will not store any data.
```

## Separating compute

> Now that you have storage restricted to specific zones, the machines in the remaining zones can be used for query processing. All you need to do is ensure that your applications connect only to the nodes in zones a,b, and c. You can either configure your applications to connect just to the remaining nodes - `127.0.0.4`, `127.0.0.5`, `127.0.0.6` or use [YugabyteDB Smart Driver](../../../drivers-orms/smart-drivers) to connect only to the nodes in zones d, e, and f using the connection string option [topology_keys](../../../drivers-orms/smart-drivers/#topology-keys) like

Reviewer: ```suggestion
Now that you have restricted storage to specific zones, the machines in the remaining zones can be used for query processing. Simply ensure that your applications connect only to the nodes in zones d, e, and f. You can either configure your applications to only connect to the remaining nodes (`127.0.0.4`, `127.0.0.5`, `127.0.0.6`), or use a [YugabyteDB Smart Driver](../../../drivers-orms/smart-drivers) to connect only to the nodes in zones d, e, and f using [topology_keys](../../../drivers-orms/smart-drivers/#topology-keys). For example:
```

```java
topology_keys = "topology_keys=aws.east.zone-d,aws.east.zone-e,aws.east.zone-f";
conn_str = "jdbc:yugabytedb://localhost:5433/yugabyte?load-balance=true&" + topology_keys;
```

## Decoupled cluster

> You have now effectively divided your cluster into 2 groups of nodes with distinct responsibilities. Your setup should look like the following illustration.  Nodes 4,5, & 6 process queries, and perform sorts and joins which are compute-heavy while nodes 1,2,3 just manage the storage of table data.

Reviewer: ```suggestion
You have now effectively divided your cluster into two groups of nodes with distinct responsibilities. Nodes 4, 5, and 6 process queries, and perform sorts and joins which are compute-heavy, while nodes 1, 2, and 3 manage the storage of table data. Your setup should look like the following illustration.
```

![Decoupled cluster](/images/explore/decoupling-compute-storage-final.png)

> By separating Compute and Storage, you can build more scalable, resilient, and efficient systems on top of YugabyteDB that can adapt to changing data and workload requirements over time.

Reviewer: ```suggestion
By separating compute and storage resources, you can build more scalable, resilient, and efficient systems on top of YugabyteDB that can adapt to changing data and workload requirements over time.
```
